tensornetwork/linalg/operations.py

```diff
@@ -0,0 +1,259 @@
+  Args:
+    a: A Tensor.
+    b: Another Tensor.
+    axes: Two lists of integers. These values are the contraction
+          axes.
+  Raises:
+    ValueError, if a and b have different backends.
+  Returns:
+    The result of the tensordot, a Tensor.
+  """
+  if a.backend.name != b.backend.name:
+    errstr = "Tried to Tensordot Tensors with differing backends \n"
+    errstr += a.backend.name + "and " + b.backend.name + "."
+    raise ValueError(errstr)
+  out_array = a.backend.tensordot(a.array, b.array, axes)
+  out_tensor = Tensor(out_array, backend=a.backend)
+  return out_tensor
+
+
+def reshape(tensor: Tensor, new_shape: Sequence[int]) -> Tensor:
+  """Reshape Tensor to the given shape.
+
+  Args:
+    tensor: Tensor to reshape.
+    new_shape: The new shape.
+  Returns:
+    The reshaped Tensor.
+  """
+  return tensor.reshape(new_shape)
+
+
+def transpose(tensor: Tensor, perm: Optional[Sequence[int]] = None) -> Tensor:
+  """ Return a new `Tensor` transposed according to the permutation set
+  by `axes`. By default the axes are reversed.
+  Args:
+    axes: The permutation. If None (default) the index order is reversed.
+  Returns:
+    The transposed `Tensor`.
+  """
+  return tensor.transpose(perm=perm)
+
+
+def take_slice(tensor: Tensor, start_indices: Tuple[int, ...],
+               slice_sizes: Tuple[int, ...]) -> Tensor:
+  """Obtains a slice of a Tensor based on start_indices and slice_sizes.
+
+  Args:
+    Tensor: A Tensor.
+    start_indices: Tuple of integers denoting start indices of slice.
+    slice_sizes: Tuple of integers denoting size of slice along each axis.
+  Returns:
+    The slice, a Tensor.
+  """
+  sliced = tensor.backend.slice(tensor.array, start_indices, slice_sizes)
+  sliced_tensor = Tensor(sliced, backend=tensor.backend)
+  return sliced_tensor
+
+
+def concatenate(tensors: Sequence[Tensor],
+                axis: Optional[int] = 0) -> Tensor:
+  """Concatenate a sequence of Tensors together about the given axis.
+  Args:
+    tensors: The list of Tensors to concatenate.
+    axis   : The axis to concatenate on. Default 0.
+  Returns:
+    The concatenated Tensor.
+  """
+  if len(tensors) <= 1:
+    raise ValueError("Must supply at least two Tensors to concatenate.")
+  all_backends_same, errstr = _check_backends(tensors, "concatenate")
+  if not all_backends_same:
+    raise ValueError(errstr)
+  arrays = [tensor.array for tensor in tensors]
+  backend = tensors[0].backend
+  concat_array = backend.shape_concat(arrays, axis)
+  return Tensor(concat_array, backend=backend)
+
+
+def shape(tensor: Tensor) -> Tuple[Optional[int], ...]:
+  """Get the shape of a Tensor as a tuple of integers.
+
+  Args:
+    Tensor: A Tensor.
+  Returns:
+    The shape of the input Tensor.
+  """
+  return tensor.shape
+
+
+def prod(values: Tensor) -> Tensor:
+  """Take the product of all of the elements in values"""
+  out_array = values.backend.shape_prod(values.array)
+  return Tensor(out_array, backend=values.backend)
```

Comment:
`shape_prod` is similar to `shape_concat` (see pytorch_backend.py)

Reply:
In this case I would propose renaming `shape_prod` to `size`

Reply:
SGTM

Reply:
Actually though, the pytorch backend code I originally modified did `np.prod(np.array(values))`, which I guess is supposed to calculate the product of a size tuple that you feed it. `Tensors` in the new implementation however store a `size` attribute which renders this redundant. It would perhaps make more sense for the ABC to just do 

`def size(tensor: Tensor) -> int:
     """Returns the total number of elements in tensor"""
     return tensor.size
`

Reply:
On reflection my previous comment did not make sense. Given the way the function operates it should probably still be called `shape_prod`, since it indeed takes products of shapes. Do you know when this is useful? It seems odd to me that there would be a special function to convert shapes into sizes, rather than just having a means to get sizes directly from backend arrays.

Reply:
I forgot why we have this. I think it is fixing some issues arising from different backends treat shapes of tensors. tensorflow for example used to have some terrible conventions there